### ISSUE TITLE
Version 1.0.1 leaks file handles

### ISSUE THREAD

**Issue by 375gnu** [kenahrens/newrelic-pm2-plugin#3]
Hi!

We use plugin version 1.0.1 with bundled pm2 version 0.12.16.

Every 30 sec plugin connects to pm2 daemon through unix socket, with every connection a new file descriptor is created, but not closed. So file handles are exhausted after some time.

**Comment by kenahrens** [comment]
@375gnu - are you seeing the file handles associated with the PM2 process or with the pm2plugin.js? How are you running pm2plugin.js, are you using a launcher like forever or even PM2 itself? Also what happens on your machine when the file handles are exhausted?

**Comment by 375gnu** [comment]
1. these file handles are associated with the pm2plugin.js process, I've found them through /proc/*/fd/
2. we use pm2 to run pm2plugin.js and our applications
3. pm2 does not accept new connections, then pm2plugin.js dies with EMFILE, and pm2 starts to accept new connections

**Comment by hatemosphere** [comment]
@kenahrens quick update: just tested version 1.0.3 and it has the same issue. pm2 plugin error log is empty in our case.
answering your question, when the limit of file handlers is reached every application that was run via pm2 starts to die, so it's kinda crucial :) can we help you to debug it somehow?

**Comment by kenahrens** [comment]
@hatemosphere - yes version 1.0.3 didn't have a fix for this file descriptor issue. I apologize if that seemed to be the case, that was a fix for something else. I have re-created the issue and I'm working on the fix, hope to have something today.

**Comment by kenahrens** [comment]
@hatemosphere @375gnu Can you try version 1.0.4?

**Comment by hatemosphere** [comment]
@kenahrens thanks a lot for your effort, seems like it stopped leaking. i've came to the same small hack with disconnect thingy and it works fine for us. i'm not good at JS, but shouldn't it close  connection by default without adding this?

**Comment by hatemosphere** [comment]
@kenahrens oops sorry for misinforming you in a previous message. 1.0.4 still has fd leaks, just doublechecked it.

```
lsof -p *pid* | wc -l
254 
```

**Comment by kenahrens** [comment]
@hatemosphere - thanks sorry been away from my desk all day. The way I tested this was by removing the POST to New Relic and turning the time way down like 1s poll cycle instead of 30s. What I found was there are a few sockets / files that are opened at startup, but they are no longer increasing over time. Are you seeing them increasing or are they staying at 254?

And regarding the need to explicitly disconnect, it wasn't clear from the PM2 docs that was required. It was some trial-and-error that got me to that approach. If you have a good source for PM2 docs please do share.  :)

**Comment by hatemosphere** [comment]
@kenahrens yep, +1 every 30 seconds, just as it was before

**Comment by kenahrens** [comment]
@hatemosphere - Sorry for the delay I was out of the office last week. I have been running version 1.0.4 since I released that version, and my fd count has been flat at 27. I'll need to track down the differences, can you get me your OS / version, PM2 version and NodeJS version?

```
$ lsof -p *pid* | wc -l
27
```

**Comment by hatemosphere** [comment]
@kenahrens seems like that was my mistake. i've reinstalled plugin and restarted pm2 and leaks are gone on 1.0.4.

```
-bash-4.1$ lsof -p 2326 | wc -l
26
```

before this i've double-checked currently running version for pm.Disconnect() and it was there. i guess some magic happened there. i think we can close this issue, many thanks to you ^_^

**Comment by kenahrens** [comment]
@hatemosphere Glad that you were able to re-test and confirm the fix in 1.0.4.
